File: julia/tests/test_conv.jl
#################################################################
#
# 1D convolution algorithms
#
#################################################################

using LinearAlgebra
#using Plots
using TickTock
using Printf
using DSP

using conv_p_m
using fftconv_p_m

L1 = 1000;
L2 = 400;
N = 50;
X = randn(L1,N);
Y = randn(L2,N);

#
# Serial Time-domain
#

Z0 = zeros(L1+L2-1,N);
tick()
for n=1:N
  Z0[:,n] = conv( X[:,n], Y[:,n]);
end
t0= tok();
@printf("conv: %f [s]\n\n", t0);

#
# Threaded time-domain
#

tick()
Z1 = conv_p_m.conv_p(X, Y)
t1 = tok()
@printf("conv_p: %f [s]\n", t1)
@printf("Numerical error: ||conv-conv_p|| = %e\n\n", sum(sum((Z0-Z1).^2)))

#
# Threaded frequency-domain
#

tick()
Z2 = fftconv_p_m.fftconv_p(X, Y)
t2 = tok()
@printf("fftconv_p: %f [s]\n", t2)
@printf("Numerical error: ||conv-fftconv_p|| = %e\n\n", sum(sum((Z0-Z2).^2)))

File: python/tests/test_conv.py
#!/usr/bin/env python

#################################################################
#
# 1D convolution algorithms
#
#################################################################

import math
import numpy as np
import scipy as sp
#import matplotlib.pyplot as plt
import time

#import sys
#sys.path.insert(0, '../')

import fftconv_p as ft_p

L1 = 1000
L2 = 400
N = 50
X = np.random.randn(L1, N)
Y = np.random.randn(L2, N)

#
# Serial Time-domain
#

Z0 = np.zeros((L1+L2-1, N))
t = time.time()
for n in range(N):
    Z0[:,n] = sp.signal.convolve(X[:,n], Y[:,n])
t0 = time.time() - t
print("conv: %f [s]" % (t0))

#
# Threaded frequency-domain
#

# NB. The first call to ft_p.fftconv_p(X, Y) is slow.
t = time.time()
Z2 = ft_p.fftconv_p(X, Y)
t2 = time.time() - t
print("fftconv_p: %f [s]\n" %  (t2))
sse = np.sum( (Z0-Z2)**2)
print("Numerical error: ||conv-fftconv_p|| = %e\n" % (sse))
